### PR TITLE
Accelerated callbacks for integration routines

### PR DESCRIPTION
**Pull request by mdewing**
This builds on #10451 

scipy.integrate routines accept compiled callbacks. This compiles Sympy expressions to a function with the appropriate signature.

The signature check on scipy.integrate unfortunately does not quite match the actual function (requires c_double rather than pointer to c_double).  So two versions of the ctypes wrapper can be generated - one that passes the signature check ('scipy.integrate'), and one that matches the actual function, which is necessary for direct calls to the function for testing ('scipy.integrate.test')

One limiting performance factor with `nquad` for multiple dimensional integration is that it calls back through python for each dimension (since it makes iterated calls to the 1-D quad routine).  Other integration packages may be better.  The `cubature` library has a python wrapper: https://github.com/saullocastro/cubature

Should this functionality be better integrated into autowrap (so other backends can provide it too)?

Some timing results with a 3-D integrand:

nquad w/python callback, time = 0.122 s
nquad w/jit callback, time = 0.048 s, speedup = 2.55
cubature w/python callback, time = 0.864 s
cubature w/jit callback, time = 0.005 s, speedup = 166.11

**Example Code**

``` python
from __future__ import print_function

import sympy.printing.llvmjitcode as jit
from sympy import exp
from sympy.abc import x,y,z
from scipy.integrate import nquad
import math
from cubature import cubature
from timeit import timeit


# Sympy expression
e = exp(-2.0*(x*x + y*y + z*z))

# Python version (for nquad)
def f(x,y,z):
    return math.exp(-2.0*(x*x + y*y + z*z))

# Python version (for cubature)
def cf(array):
    x = array[0]
    y = array[1]
    z = array[2]
    return math.exp(-2.0*(x*x + y*y + z*z))

ndim = 3
lim = [0.0, 10.0]
lims = [lim]*ndim
lim_min = [lim[0]]*ndim
lim_max = [lim[1]]*ndim

# Scipy.integrate.nquad

# Python callback
def run_nquad():
    nquad(f, lims)
nquad_time = timeit(stmt=run_nquad, number=1)
print('nquad w/python callback, time = %.3f s'%nquad_time)

# JIT compiled callback
e_scipy = jit.llvm_callable([x,y,z], e, callback_type='scipy.integrate')
def run_nquad_jit():
    nquad(e_scipy, lims)
nquad_time_jit = timeit(stmt=run_nquad_jit, number=1)
print('nquad w/jit callback, time = %.3f s, speedup = %.2f'%(nquad_time_jit,nquad_time/nquad_time_jit))


# Cubature

# Python callback
def run_cubature():
    cubature(cf, ndim=3, fdim=1, xmin=lim_min, xmax=lim_max)
cubature_time = timeit(stmt=run_cubature, number=1)
print('cubature w/python callback, time = %.3f s'%cubature_time)


# JIT compiled callback
e_cub = jit.llvm_callable([x,y,z], e, callback_type='cubature')
def run_cubature_jit():
    cubature(e_cub, ndim=3, fdim=1, xmin=lim_min, xmax=lim_max)
cubature_time_jit = timeit(stmt=run_cubature_jit, number=1)
print('cubature w/jit callback, time = %.3f s, speedup = %.2f'%(cubature_time_jit,cubature_time/cubature_time_jit))
```

### CODE REVIEW

**Comment by asmeurer** [comment]
Relevant https://stackoverflow.com/questions/35430479/convert-sympy-expressions-to-function-of-numpy-arrays

**Comment by asmeurer** [comment]
I take it https://github.com/sympy/sympy/pull/10451 should be merged first.

**Comment by markdewing** [comment]
Thanks for the pointer to the Stack Overflow question. There are two steps that could speed performance
1. Compile to code that accepts and returns numpy arrays.   The evaluation of the function is faster, but there is still the overhead of the scipy wrapper code around the underlying ODE integrator.
2. Modify scipy to accept raw callbacks (as was done for `quad`), and then compile to code for that interface.

**Comment by asmeurer** [comment]
Is this ready to be reviewed/merged or are more changes coming?

**Comment by markdewing** [comment]
It should be ready to go. The changes for addressing the stack overflow issue you pointed to are more involved to add here, and would go in a separate PR.

File: sympy/printing/llvmjitcode.py [hunk 1]
```diff
@@ -220,16 +318,66 @@ def llvm_callable(args, expr):
...
+                      ctypes.POINTER(ctypes.c_double),
+                      ctypes.c_void_p,
+                      ctypes.c_int,
+                      ctypes.POINTER(ctypes.c_double)
+                      ]
+        signature.ret_type = ctypes.c_int
+        signature.input_arg = 1
+        signature.ret_arg = 4
+    else:
+        print("Unknown callback type: %s" % callback_type)
```
Comment (asmeurer) [inline sympy/printing/llvmjitcode.py]: `raise ValueError("Unknown callback type: %s" % callback_type)`

**Comment by markdewing** [comment]
Here's an example using Sympy plus these integration callbacks to compute the ground state energy of a  hydrogen atom.  It also displays some of the performance of various callbacks/integration packages.

https://github.com/markdewing/next_steps_in_programming/blob/master/examples/integration/Hydogen%20Atom.ipynb

**Pull request merged**